annet/mesh/basemodel.py

```diff
@@ -0,0 +1,155 @@
+from abc import ABC
+from copy import copy
+from enum import Enum
+from typing import TypeVar, Any, Annotated, get_origin, get_type_hints, get_args, Callable
+
+
+class MergeForbiddenError(Exception):
+    pass
+
+
+class Special(Enum):
+    NOT_SET = "<NOT SET>"
+
+
+T = TypeVar('T')
+
+
+class Merger(ABC):
+    def __call__(self, name: str, x: T | Special, y: T | Special) -> T:
+        if x is Special.NOT_SET:
+            return y
+        if y is Special.NOT_SET:
+            return x
+        return self._merge(name, x, y)
+
+    def _merge(self, name: str, x: T, y: T) -> T:
+        raise NotImplementedError
+
+
+class UseFirst(Merger):
+    def _merge(self, name: str, x: T, y: T) -> T:
+        return x
+
+
+class UseLast(Merger):
+    def _merge(self, name: str, x: T, y: T) -> T:
+        return y
+
+
+class Forbid(Merger):
+    def _merge(self, name: str, x: T, y: T) -> T:
+        raise MergeForbiddenError(f"Override is forbidden for field {name}")
+
+
+class ForbidChange(Merger):
+    def _merge(self, name: str, x: T, y: T) -> T:
+        if x == y:
+            return x
+        raise MergeForbiddenError(
+            f"Override with different value is forbidden for field {name}:\n"
+            f"Old: {x}\n"
+            f"New: {y}"
+        )
+
+
+class Concat(Merger):
+    def _merge(self, name: str, x: T, y: T) -> T:
+        return x + y
```

Comment:
It is not stated what T implements `__add__`
```
class T(Protocol):
    def __add__(self, a: 'T') -> 'T':
        pass
```

Reply:
I am not sure there is any profit of specifying constraints on `TypeVar` here as it is not validated by linter

Reply:
The whole point of annotations is to satisfy linters in IDE and tests, generate docs and make users' life easier. Maybe broken annotations are better than no annotations, but I think we should move towards of annotations validation in tests.
BTW my vscode already highlight this discrepancy.

Reply:
I understand your point, by the case to validate here is:
```py
a: Annotated[A, Concat()]
```

No tool will check that A supports `__add__`